### PR TITLE
[SPARK-16672][SQL] SQLBuilder should not raise exceptions on EXISTS queries

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
## What changes were proposed in this pull request?

Currently, `SQLBuilder` raises `empty.reduceLeft` exceptions on _unoptimized_ `EXISTS` queries. We had better prevent this.

``` scala
scala> sql("CREATE TABLE t1(a int)")
scala> val df = sql("select * from t1 b where exists (select * from t1 a)")
scala> new org.apache.spark.sql.catalyst.SQLBuilder(df).toSQL
java.lang.UnsupportedOperationException: empty.reduceLeft
```
## How was this patch tested?

Pass the Jenkins tests with a new test suite.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Hi, @hvanhovell .
Could you review this PR about `SQLBuilder`?

**Comment by jaceklaskowski** [comment]
I didn't even know such a SQL query is possible in SQL :) Lots to learn still.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/SQLBuilderSuite.scala [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ */
+
+package org.apache.spark.sql.hive
+
+import org.apache.spark.sql.QueryTest
+import org.apache.spark.sql.catalyst.SQLBuilder
+import org.apache.spark.sql.hive.test.TestHiveSingleton
+import org.apache.spark.sql.test.SQLTestUtils
+
+class SQLBuilderSuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
```
Comment (rxin) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/SQLBuilderSuite.scala]: LogicalPlanToSQLSuite?

Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/SQLBuilderSuite.scala]: Oh, I see.

**Comment by dongjoon-hyun** [comment]
Rebased.

**Comment by dongjoon-hyun** [comment]
Hi, @hvanhovell .
Could you review this PR?

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/SQLBuilder.scala [hunk 2]
```diff
@@ -512,8 +512,12 @@ class SQLBuilder(logicalPlan: LogicalPlan) extends Logging {
           ScalarSubquery(rewrite, Seq.empty, exprId)
 
         case PredicateSubquery(query, conditions, false, exprId) =>
-          val plan = Project(Seq(Alias(Literal(1), "1")()),
-            Filter(conditions.reduce(And), addSubqueryIfNeeded(query)))
+          val plan = if (conditions.isEmpty) {
+            addSubqueryIfNeeded(query)
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/SQLBuilder.scala]: Minor: it is a little more readable if we move the `addSubqueryIfNeeded(query)` out of the if...else... and into a separate variable.

**Comment by hvanhovell** [comment]
One minor comment. Looks pretty good.

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/SQLBuilder.scala (hunk 2 above)
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/SQLBuilder.scala]: Oh, right. I'll fix soon.

**Comment by dongjoon-hyun** [comment]
Thank you for review, @hvanhovell .
I updated that.

**Comment by hvanhovell** [comment]
LGMT, pending jenkins

**Comment by dongjoon-hyun** [comment]
It's strange . Jenkins does not trigger the test for the last commit.

**Comment by dongjoon-hyun** [comment]
Oh, it became a maintenance mode. :(

**Comment by dongjoon-hyun** [comment]
Hi, @hvanhovell .
Could you merge this PR? :)

File: sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala [hunk 3]
```diff
@@ -927,6 +927,14 @@ class LogicalPlanToSQLSuite extends SQLBuilderTest with SQLTestUtils {
     }
   }
 
+  test("predicate subquery") {
+    withTable("t1") {
+      sql("CREATE TABLE t1(a int)")
+      val df = sql("select * from t1 b where exists (select * from t1 a)")
```
Comment (rxin) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: it'd be good to test the generated sql.

Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: Sure. No problem.

**Comment by dongjoon-hyun** [comment]
Hi, @rxin .
The PR is updated to use `checkSQL`.

**Comment by dongjoon-hyun** [comment]
Hi, @rxin .
Could you merge this PR please?
Since this PR uses `checkSQL`, it will conflict with `Generate stable SQLs in SQLBuilder`(https://github.com/apache/spark/pull/14257)
After merging this, I will rebase to the master and regenerate the answer in https://github.com/apache/spark/pull/14257.

**Comment by rxin** [comment]
Merging in master/2.0.

**Comment by dongjoon-hyun** [comment]
Thank you so much for review and merging, @rxin , @hvanhovell , and @jaceklaskowski !

**Pull request closed without merging**